content/blog/exploring-zookeeper-less-kafka.asciidoc

```diff
@@ -0,0 +1,464 @@
+---
+title: "Exploring ZooKeeper-less Kafka"
+date: 2021-05-17T18:45:00+01:00
+draft: false
+---
+:source-highlighter: rouge
+:rouge-style: base16.dark
+:icons: font
+:imagesdir: /images
+ifdef::env-github[]
+:imagesdir: ../../static/images
+endif::[]
+
+Sometimes, less is more.
+One case where that's certainly true is dependencies.
+And so it shouldn't come at a surprise that the https://kafka.apache.org/[Apache Kafka] community is eagerly awaiting the removal of the dependency to the https://zookeeper.apache.org/[ZooKeeper] service,
+which currently is used for storing Kafka metadata (e.g. about topics and partitions) as well as leader election in the cluster.
+
+https://cwiki.apache.org/confluence/display/KAFKA/KIP-500%3A+Replace+ZooKeeper+with+a+Self-Managed+Metadata+Quorum[KIP-500]
+("Replace ZooKeeper with a Self-Managed Metadata Quorum")
+promises to make life better for users in many regards:
+
+* Better getting started and operational experience by requiring to run only one system, Kafka, instead of two
+* Removing potential for discrepancies of metadata state between ZooKeeper and the Kafka controller
+* Simplifying configuration, for instance when it comes to security
+* Better scalability, e.g. in terms of number of partitions, faster execution of operations like topic creation
+
+<!--more-->
+
+With KIP-500, Kafka itself will store all the required metadata in internal topics,
+and controller election will be done amongst (a subset of) the Kafka cluster nodes themselves.
+
+Having been in the works for multiple years, ZK-less Kafka,
+also known as KRaft ("Kafka Raft metadata mode"), was recently published as an early access feature with https://blogs.apache.org/kafka/entry/what-s-new-in-apache5[Kafka 2.8].
+I.e. the perfect time to get my hands on this and get a first feeling for ZK-less Kafka myself.
+Note this post isn't meant to be a thorough evaluation or systematic testing of the new Kafka deployment mode,
+rather take it as a description of how to get started playing with ZK-less Kafka and of a few observations I made while doing so.
+
+In the world of ZK-less Kafka, there's two node roles for nodes: _controller_ and _broker_.
+Each node in the cluster can have either one or both roles ("combined nodes").
+All controller nodes elect the _active controller_ (sometimes also simply referred to as _leader_),
```

Comment:
I think controller is different from leader? Leader is the primary responsibility for a partition, while the controller is the primary for the Kafka cluster coordination? Or did that change?

Reply:
Yeah, so I was confused about this initially, too. But if you look at the different KIPs, "leader" is used there for referring to the active controller, so it's a bit ambiguous. But then the active controller also is the partition leader of the single metadata topic partition, so it's also consistent in a way :)

Reply:
Expanded a bit on this in the post.